change deb repos

Spirit-act opened the pull request:
## Description

RabbitMQ changed the Debian / Ubuntu repos again.

I have:
- change the repo to the new urls
- changed the signing keys accordingly
- incremented the version to new current release
- fixed an ansible boolean comparison (relevant for current and future ansible releases)

## Info

I have not adjusted the RPM release because I have no option to test it and can therefor not confirm if it works.

## Breaking Changes

- Default RabbitMQ version updated from 3.13.7 to 4.2.1 (major version)
- Users should override `rabbitmq_debian_version` to stay on 3.x if not ready to upgrade
- See RabbitMQ 4.0 Release Notes for upgrade considerations

## Testing

- [x] Tested on Ubuntu 20.04 (molecule) > ! The old Python3 version does no longer work with current ansible [1]
- [x] Tested on Ubuntu 22.04 (molecule)
- [x] Tested on Ubuntu 24.04 (molecule + prod system)
- [ ] Tested on Debian 11
- [ ] Tested on Debian 12

## requirements.txt

I have updated the requirements.txt to use the current ansible version.

[1] The current ansible versions require python 3.9 or newer. Ubuntu 20.04 has only python 3.8.

Spirit-act committed:
change deb packages to new repo

Signed-off-by: Spirit-act <16631523+Spirit-act@users.noreply.github.com>

Spirit-act committed:
adjust boolean check for future ansible versions

Signed-off-by: Spirit-act <16631523+Spirit-act@users.noreply.github.com>

mrlesmithjr reviewed (changes requested):
## Code Review: REQUEST CHANGES

Thank you for addressing the RabbitMQ repository URL changes! However, there are several critical issues that need attention before this can be merged.

---

### Critical Issues

#### 1. Breaking Version Upgrade (3.13.7 -> 4.2.1)

**Location:** `defaults/main.yml:42`

```yaml
-rabbitmq_debian_version: 3.13.7-1
+rabbitmq_debian_version: 4.2.1
```

**Problem:** This is a **major version upgrade** that could break existing deployments. RabbitMQ 4.x has different features and potentially breaking changes.

**Impact:**
- Existing users will unexpectedly upgrade to a major version
- No upgrade path documentation provided
- Could cause production outages for existing clusters

**Suggested:**
- Keep default at latest 3.x version for backward compatibility, OR
- Document this as a breaking change with upgrade instructions, OR
- Add a clear warning in the PR description

---

#### 2. Repository URL Pattern Needs Verification

**Location:** `defaults/main.yml:31`

```yaml
rabbitmq_debian_repo: "deb [...] https://deb1.rabbitmq.com/rabbitmq-server/{{ ansible_facts['distribution'] | lower }}/{{ ansible_facts['distribution_release'] }} {{ ansible_facts['distribution_release'] }} main"
```

**Concern:** The URL includes `{{ ansible_facts['distribution_release'] }}` twice - once in the path and once as the suite name. This pattern seems unusual.

**Questions:**
- Does `https://deb1.rabbitmq.com/rabbitmq-server/ubuntu/jammy/jammy/main/` actually exist?
- Has this been tested on both Ubuntu and Debian?

**Verification needed:**
```bash
curl -I https://deb1.rabbitmq.com/rabbitmq-server/ubuntu/jammy/
```

---

#### 3. Version Format Changed

**Location:** `defaults/main.yml:42`

```yaml
-rabbitmq_debian_version: 3.13.7-1
+rabbitmq_debian_version: 4.2.1
```

The version format changed from `X.Y.Z-REVISION` to `X.Y.Z`. If the package in the new repo is named `rabbitmq-server=4.2.1-1`, installation will fail.

**Please verify:** The exact package version string in the new repository.

---

### Major Issues

#### 4. Inconsistent Ansible Facts Format

**Location:** `defaults/main.yml:31,36`

The PR introduces `ansible_facts['distribution']` but other parts of the role use `ansible_distribution_release`. While `ansible_facts` is the modern format, consistency is important.

---

#### 5. Molecule Tests Outdated

**Location:** `molecule/ubuntu2004/converge.yml:6`

```yaml
rabbitmq_debian_repo: "deb https://dl.bintray.com/rabbitmq-erlang/debian..."
```

Bintray was shut down in 2021. These tests will fail and should be updated.

---

### Positive Aspects

1. **Good:** Boolean check improvement from `when: rabbitmq_plugins` to `when: rabbitmq_plugins | length > 0`
2. **Good:** Proactively updating to new repository infrastructure

---

### Requested Actions

1. **Revert version to 3.x** or clearly document the breaking change
2. **Verify repository URL structure** with actual curl tests
3. **Test on Ubuntu and Debian** systems
4. **Update molecule tests** to use new repository URLs
5. **Add testing evidence** to PR description

---

### Documentation Suggestion

If keeping 4.x, add to PR description:
```markdown
## Breaking Changes

- Default RabbitMQ version updated from 3.13.7 to 4.2.1 (major version)
- Users should override `rabbitmq_debian_version` to stay on 3.x if not ready to upgrade
- See RabbitMQ 4.0 Release Notes for upgrade considerations

## Testing

- [ ] Tested on Ubuntu 20.04
- [ ] Tested on Ubuntu 22.04
- [ ] Tested on Debian 11
- [ ] Tested on Debian 12
```

---

Looking forward to the updates!

Spirit-act commented:
Thank you for your review.

I have some questions and clarifications.

### 1 and 3. Version Change

I would prefer not to revert back to version 3. If somebody wants to deploy a rabbitmq installation or cluster i would assume this person wants the latest version.

I also did not thought about the revision numbers. I will adjust this.

### 2.  Repo URL

There a two different varaibles used:

Distribution:
```
ansible_facts['distribution']
```
and
Distribution release:
```
ansible_facts['distribution_release']
```

This is needed, because the repo differents between Ubuntu and Debian like:
```
https://deb1.rabbitmq.com/rabbitmq-erlang/ubuntu/noble noble main

https://deb1.rabbitmq.com/rabbitmq-server/debian/trixie trixie main
```

Source: https://www.rabbitmq.com/docs/install-debian

### 4. Inconsistency in fact variables

I will change all variables to the array format.

Because in the porting guide for Ansible 13 the following format is deprecated as default:
```
ansible_*
```

Source: https://docs.ansible.com/projects/ansible/latest/porting_guides/porting_guide_13.html#inject-facts-as-vars 

### 5: Molecule tests

I currently do not have any experience with molecule, but I'll try to add the requested tests.

### Test Coverage

We are using the role with the proposed changes on Ubuntu 24.04 LTS.

Spirit-act committed:
add revision to rabbitmq_debian_version

Signed-off-by: Spirit-act <16631523+Spirit-act@users.noreply.github.com>

Spirit-act committed:
use consistent ansible_facts variables format

Signed-off-by: Spirit-act <16631523+Spirit-act@users.noreply.github.com>

Spirit-act committed:
add molecule test for 24.04 and 22.04

- add molecule files for ubuntu 22.04 and 24.04
- disable ubuntu 18.04 molecule test
- update workflow action/cache to version 4

Signed-off-by: Spirit-act <16631523+Spirit-act@users.noreply.github.com>

Spirit-act commented:
## RedHat

After running `ansible-lint` I changed the fqcn of the package management tasks from `yum` to `dnf`.  [Source](https://docs.ansible.com/projects/ansible/latest/collections/ansible/builtin/yum_module.html)


## requirments

In the following commit ([00c8291](https://github.com/mrlesmithjr/ansible-rabbitmq/pull/76/commits/00c8291b33205e92d776b71129878ed853ae58b7)) I updated the `requirments.txt` and the `requirments-dev.txt` to use the current ansible version.

I used the `pip freeze` command to generate the list. The command did not add the python_full_version informations.

The following python packages where not in my test environment. Maybe something changed or there were unnecessary packages in the `requirements-dev.txt`.

Please have a look at it:
```text
arrow==1.3.0 ; python_full_version >= "3.8.1" and python_version < "4.0"     
astroid==2.15.8 ; python_full_version >= "3.8.1" and python_version < "4.0"
binaryornot==0.4.4 ; python_full_version >= "3.8.1" and python_version < "4.0"
chardet==5.2.0 ; python_full_version >= "3.8.1" and python_version < "4.0"
colorama==0.4.6 ; python_full_version >= "3.8.1" and python_version < "4.0" and (sys_platform == "win32" or platform_system == "Windows")
cookiecutter==2.5.0 ; python_full_version >= "3.8.1" and python_version < "4.0"
dill==0.3.7 ; python_full_version >= "3.8.1" and python_version < "4.0"
importlib-resources==6.1.1 ; python_full_version >= "3.8.1" and python_version < "3.9"
isort==5.13.2 ; python_full_version >= "3.8.1" and python_version < "4.0"
lazy-object-proxy==1.10.0 ; python_full_version >= "3.8.1" and python_version < "4.0"
pkgutil-resolve-name==1.3.10 ; python_full_version >= "3.8.1" and python_version < "3.9"
pylint==2.17.7 ; python_full_version >= "3.8.1" and python_version < "4.0"
python-dateutil==2.8.2 ; python_full_version >= "3.8.1" and python_version < "4.0"
python-slugify==8.0.1 ; python_full_version >= "3.8.1" and python_version < "4.0"
pywin32==306 ; python_full_version >= "3.8.1" and python_version < "4.0" and sys_platform == "win32"
setuptools==69.0.2 ; python_full_version >= "3.8.1" and python_version < "4.0"
six==1.16.0 ; python_full_version >= "3.8.1" and python_version < "4.0"
text-unidecode==1.3 ; python_full_version >= "3.8.1" and python_version < "4.0"
tomli==2.0.1 ; python_full_version >= "3.8.1" and python_full_version < "3.11.0a7"
tomlkit==0.12.3 ; python_full_version >= "3.8.1" and python_version < "4.0"
types-python-dateutil==2.8.19.14 ; python_full_version >= "3.8.1" and python_version < "4.0"
typing-extensions==4.9.0 ; python_full_version >= "3.8.1" and python_version < "3.11"
wrapt==1.16.0 ; python_full_version >= "3.8.1" and python_version < "4.0"
zipp==3.17.0 ; python_full_version >= "3.8.1" and python_version < "3.9"
```

Spirit-act committed:
update requirements to use current ansible version

Signed-off-by: Spirit-act <16631523+Spirit-act@users.noreply.github.com>

Spirit-act committed:
update multiple pipeline breaking packages

- update python3 version to 3.13
- update versions in pre-commit-config
- remove trailing empty line in requirements.txt
- adjust facts variables in main.yml
- ubuntu20.04 has python3.8 which is no longer supported by current
  ansible versions

Signed-off-by: Spirit-act <16631523+Spirit-act@users.noreply.github.com>

mrlesmithjr committed:
fix: remove arch=amd64 restriction for multi-arch support

- Remove hardcoded arch=amd64 from rabbitmq_debian_repo and
  rabbitmq_debian_erlang_repo to support ARM64 systems
- Update GitHub Actions: checkout@v3→v4, setup-python@v4→v5

Note: RabbitMQ server ARM64 packages are now accessible. However,
full ARM64 installation may still require providing Erlang 26+ from
an alternative source, as RabbitMQ's Erlang repository does not
currently provide ARM64 builds.

mrlesmithjr commented:
## Maintainer Update

Thanks for updating the repository URLs! I've pushed a small fix to this branch:

### Changes Made

1. **Removed `arch=amd64` restriction** from both `rabbitmq_debian_repo` and `rabbitmq_debian_erlang_repo` in `defaults/main.yml`
   - This allows the role to work on ARM64 systems (Apple Silicon, AWS Graviton, etc.)
   
2. **Updated GitHub Actions versions** in `.github/workflows/default.yml`
   - `actions/checkout@v3` → `@v4`
   - `actions/setup-python@v4` → `@v5`

### ARM64 Limitation (FYI)

During testing, I discovered that while RabbitMQ **server** packages are available for ARM64 from the new repos, RabbitMQ's **Erlang repository** does not provide ARM64 builds. This means:

| Architecture | Status |
|--------------|--------|
| AMD64/x86_64 | ✅ Full install works |
| ARM64 | ⚠️ RabbitMQ packages accessible, but Erlang 26+ must come from another source |

This is an upstream RabbitMQ packaging decision, not something we can fix in this role. ARM64 users can work around this by:
- Setting `rabbitmq_debian_erlang_from_rabbit: false` and using system Erlang
- Using Ubuntu 24.04+ which may have newer Erlang in default repos
- Installing Erlang from another source

---

The PR is ready to merge once CI passes. 🚀

Spirit-act commented:
@mrlesmithjr I think this should be ready for merging, before it goes stale.
